Add Inngest guide (#645)
clerk/clerk-docs: docs/integrations/webhooks/inngest.mdx

---
title: Handling webhooks with Inngest
description: Learn how to integrate Clerk with Inngest.
---

# Handling webhooks with Inngest

> Webhooks allow you to [synchronize data](/docs/users/sync-data) from Clerk to your application backend. You can either handle them directly in your backend with an endpoint or use a tool like [Inngest](https://inngest.com/) which receives the webhook events for you and reliably executes functions in your codebase. When handling webhooks, Inngest receives the webhook events for you and uses a built-in queue to reliably execute longer running functions with additional functionality including:

Reviewer: ```suggestion
Webhooks allow you to [synchronize data](/docs/users/sync-data) from Clerk to your application backend. You can either handle them directly in your backend with an endpoint or use a tool like [Inngest](https://inngest.com/) which receives the webhook events for you and reliably executes functions in your codebase. When handling webhooks, Inngest receives the [webhook events](/docs/integrations/webhooks#supported-webhook-events) for you and uses a built-in queue to reliably execute longer running functions with additional functionality including:
```

* [Limiting concurrency](https://www.inngest.com/docs/guides/concurrency) to handle spikes in events without overwhelming your API or database.
* Triggering multiple functions from a single event ([fan-out jobs](https://www.inngest.com/docs/guides/fan-out-jobs)).
* [Delaying code](https://www.inngest.com/docs/functions/multi-step) to run after a period of time.
* [Debouncing events](https://www.inngest.com/docs/reference/functions/debounce) to reduce duplicate processing.

In this guide, you'll learn how to set up Inngest to receive Clerk webhook events and how to define Inngest functions in your application using Clerk events including synchronizing data and sending welcome emails.

> To follow this guide, you need an Inngest account and have Inngest [set up](https://www.inngest.com/docs/quick-start) in your codebase.

Reviewer 2: ```suggestion
To follow this guide, you need an Inngest account (free tier is enough) and have Inngest [set up](https://www.inngest.com/docs/quick-start) in your codebase.
```

## Setting up the Inngest webhook

To create an Inngest webhook endpoint and add it to your Clerk account, open the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) page at the Clerk dashboard. Next, select the **Add Endpoint** button.

<Images
  width={2578}
  height={1652}
  src="/docs/images/integrations/inngest/webhook-page.webp"
  alt="The Webhooks page in the Clerk Dashboard. A red arrow points to the button for Add Endpoint."
/>

On the next page, select the **Transformation template** tab and the **Inngest** template, then click on the **Connect to Inngest** button.

<Images
  width={1906}
  height={1376}
  src="/docs/images/integrations/inngest/webhook-transformation-template.webp"
  alt="The Webhooks page in the Clerk Dashboard showing the Inngest transformation template. Red arrows point to the Transformation Template tab, the Inngest template, and the Connect to Inngest button."
/>

A popup window will appear to complete the setup. Select **Approve** to create the webhook.

> <Images
>   width={1880}
>   height={1448}
>   src="/docs/images/integrations/inngest/webhook-endpoint-connected.webp"
>   alt="The Webhooks page in the Clerk Dashboard showing a connected Inngest account. A red arrow points to the Connected button."
> />

Reviewer: This image appears to be in the wrong order. Looks like it should be flipped with the one below it. 
```suggestion
<Images
  width={1194}
  height={1520}
  src="/docs/images/integrations/inngest/inngest-permissions-dialog.webp"
  alt="The Inngest permissions popup window showing the Approve button."
/>
```

After the popup window disappears, the Webhooks page will now display **Connected** with the webhook URL underneath. There is one more step to complete setup.

> <Images
>   width={1194}
>   height={1520}
>   src="/docs/images/integrations/inngest/inngest-permissions-dialog.webp"
>   alt="The Inngest permissions popup window showing the Approve button."
> />

Reviewer: ```suggestion
<Images
  width={1880}
  height={1448}
  src="/docs/images/integrations/inngest/webhook-endpoint-connected.webp"
  alt="The Webhooks page in the Clerk Dashboard showing a connected Inngest account. A red arrow points to the Connected button."
/>
```

To complete the setup, scroll down and select **Create**.

<Images
  width={1906}
  height={1376}
  src="/docs/images/integrations/inngest/webhook-create.webp"
  alt="The Webhooks page in the Clerk Dashboard showing the end of the page to create a new endpoint. A red arrow points to the Create button."
/>

You'll be redirected to the new endpoint. In your Inngest dashboard, you will see a new webhook created in your account's [production environment](https://app.inngest.com/env/production/manage/webhooks).

## Viewing webhook events within Inngest

> After setup, as webhook events are sent from Clerk to Inngest, new `clerk `events will appear  in your Inngest dashboard. Event names will be the prefixed `clerk/` followed by the event name. See [webhook events](/docs/integrations/webhooks#supported-webhook-events) for a full list.

Reviewer: Maybe link 'Inngest dashboard' to the dashboard so it is easy to view?

> After setup, as webhook events are sent from Clerk to Inngest, new `clerk `events will appear  in your Inngest dashboard. Event names will be the prefixed `clerk/` followed by the event name. See [webhook events](/docs/integrations/webhooks#supported-webhook-events) for a full list.

Reviewer 2: ```suggestion
After setup, as webhook events are sent from Clerk to Inngest, new `clerk` events will appear in your [Inngest dashboard](https://app.inngest.com/env/production/events). Event names will be the prefixed `clerk/` followed by the event name. See [webhook events](/docs/integrations/webhooks#supported-webhook-events) for a full list.
```

<Images
  width={1052}
  height={860}
  src="/docs/images/integrations/inngest/inngest-clerk-events.webp"
  alt="The Events page in the Inngest Dashboard showing a list of Clerk events."
/>

## Creating a function to sync a new user to a database

With Inngest already set up in your codebase, you can now use these events as triggers for your functions.

> The code below features the `inngest.createFunction` method, which inserts a new user into the database. It will be triggered whenever `clerk/user.created` event occurs.

Reviewer 2: ```suggestion
Suppose you need to write a function which will insert a new user into the database which will be triggered whenever `clerk/user.created` event occurs. You would use the `inngest.createFunction` method, like in the example below:
```

> ```typescript

Reviewer 2: ```suggestion
```typescript filename="src/inngest/sync-user.ts"
```

const syncUser = inngest.createFunction(
 { id: 'sync-user-from-clerk' },
 { event: 'clerk/user.created' }, // ← This is the function's triggering event
 async ({ event }) => {
   const user = event.data; // The event payload's data will be the Clerk User json object
   const { id, first_name, last_name } = user;
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email
   await database.users.insert({ id, email, first_name, last_name })
 }
)

> ```

Reviewer: I think we can make this a little clearer that this in the customer's codebase. You can use something like `filename="src/inngest/sync-user.ts"` on line 88 after the language as well. 

Also, having not used Inngest and just reading this I am wondering if the 'id' value is just a label as it appears to be or if it should be an id string of some sort.

Reviewer 2: Good idea and feedback. I've changed the description paragraph, added the file name, and a comment about the `id` ✅ Thank you!

The `event` object contains all of the relevant data for the event. The `event.data` will match the `data` object from the standard Clerk webhook [payload structure](/docs/integrations/webhooks#payload-structure). With this `clerk/user.created` event, the `event.data` will be a Clerk User json object.

As you can see, you can choose which events you want to handle with each function. You might write a separate function for `clerk/user.updated` and `clerk/user.deleted` handling the entire lifecycle end to end.

Note that multiple functions can also listen to the same event. This pattern is called “[fan-out](https://www.inngest.com/docs/guides/fan-out-jobs).”

## Creating a function to send a welcome email

Often, applications need to perform additional tasks when a new user is created, like send a welcome email with tips and useful information.

> While it is possible to add this logic at the end of your sync function as seen in the [previous section](/docs/integrations/webhooks/inngest#creating-a-function-to-sync-a-new-user-to-a-database), it's better to decouple logic into different functions so issues with one task do not affect the other ones. For example, if your email fails to send, it should not affect adding the user to an email list.

Reviewer 2: ```suggestion
While it is possible to add this logic at the end of your sync function as seen in the [previous section](/docs/integrations/webhooks/inngest#creating-a-function-to-sync-a-new-user-to-a-database), it’s better to decouple unrelated tasks into different functions so issues with one task do not affect the other ones. For example, if your email fails to send, it should not affect starting a trial for that user in Stripe.
```

> While it is possible to add this logic at the end of your sync function as seen in the [previous section](/docs/integrations/webhooks/inngest#creating-a-function-to-sync-a-new-user-to-a-database), it's better to decouple logic into different functions so issues with one task do not affect the other ones. For example, if your email fails to send, it should not affect adding the user to an email list.

Reviewer 2: @royanger let me know if this paragraph is now clearer 🙏

Reviewer: Looks great! Feel free to resolve this once you see!

With Inngest, each function has automatic retries, so only the code that has issues is re-run.

The code below creates another function using the same `clerk/user.created` event and adds the logic to send the welcome email:

```typescript
const sendWelcomeEmail = inngest.createFunction(
 { id: 'send-welcome-email' },
 { event: 'clerk/user.created' },
 async ({ event }) => {
   const { user } = event.data
   const { first_name } = user
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email

>    await emails.sendWelcomeEmail({ email, first_name })

Reviewer: I believe that `emails` is just a placeholder?
```suggestion
   // `emails` is a placeholder for the function in your codebase to send email
   await emails.sendWelcomeEmail({ email, first_name })
```

 }
)
```

Now, you have a function that utilizes the same Clerk webhook event for another purpose. Clerk webhook events can be used for all sorts of application lifecycle use cases. For example, adding users to a marketing email list, starting a Stripe trial, or provisioning new account resources.

### Sending a delayed follow-up email

Every Inngest function handler has an additional `step` object which provides tools to create more complex functions. Using `step.run` allows you to encapsulate specific code that will be automatically retried ensuring that issues with one part of your function don't force the entire function to re-run. Additionally, [other tools like `step.sleep`](https://www.inngest.com/docs/reference/functions/step-sleep), are available to extend functionality.

> The code below sends a welcome email, then uses `step.sleep` to wait for three days before sending another email offering a free trial:

Reviewer: In the previous section the guide indicates to decouple tasks so they don't affect each other, but this example has two tasks coupled together. Is there a reason for this? It came off weird reading the guide going from the previous section to this one section.

Reviewer 2: That’s a good point. It makes sense to decouple either big tasks or unrelated tasks. Here, one task results from another and they constitute one whole.

I'll rework the previous section a little to make this clearer!

Reviewer 2: See https://github.com/clerk/clerk-docs/pull/645/files/388815e0218521951e3e4dc48ebc67989798c6a9#r1459451471

```typescript
const sendOnboardingEmails = inngest.createFunction(
 { id: 'onboarding-emails' },
 { event: 'clerk/user.created' },
 async ({ event, step }) => { // ← step is available in the handler's arguments
   const { user } = event.data
   const { first_name } = user
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email

   await step.run('welcome-email', async () => {

>      await emails.sendWelcomeEmail({ email, first_name })

Reviewer: ```suggestion
     // `emails` is a placeholder for the function in your codebase to send email
     await emails.sendWelcomeEmail({ email, first_name })
```

   })

   // wait 3 days before second email
   await step.sleep('wait-3-days', '3 days')

   await step.run('trial-offer-email', async () => {
     await emails.sendTrialOfferEmail({ email, first_name })
   })
 }
)
```

Now, you've extended the usefulness of Clerk webhook events even further to build an onboarding drip email campaign in just a few lines of code.

## Testing webhook events using the Inngest Dev Server

During local development with Inngest, you can use the [Inngest Dev Server](https://www.inngest.com/docs/local-development) to run and test your functions on your own machine. To start the server, in your project directory run the following command:

```sh filename="terminal"
npx inngest-cli@latest dev
```

In your browser open [http://localhost:8288](http://localhost:8288) to see the development UI.

> To quickly get events to test within Dev Server, you can select any individual event from the **Events** tab then select the **Send to Dev Server**.

Reviewer: It wasn't super clear here that this step was back in the Inngest dashboard.

Reviewer 2: I've just pushed a commit to the previous paragraph. Do you think it clarifies that we are now operating on the Dev Server UI? (It's not the Inngest Dashboard, it's local development)

Reviewer: This definitely better. I was confused -- I ran the dev server but didn't really do anything with it. It looked different on its starting screen so I thought the following step was back in the Dashboard. Great clarification for someone just reading!

Please resolve once you read this.

---
title: Handling webhooks with Inngest
description: Learn how to integrate Clerk with Inngest.
---

# Handling webhooks with Inngest

Webhooks allow you to [synchronize data](/docs/users/sync-data) from Clerk to your application backend. You can either handle them directly in your backend with an endpoint or use a tool like [Inngest](https://inngest.com/) which receives the webhook events for you and reliably executes functions in your codebase. When handling webhooks, Inngest receives the webhook events for you and uses a built-in queue to reliably execute longer running functions with additional functionality including:

* [Limiting concurrency](https://www.inngest.com/docs/guides/concurrency) to handle spikes in events without overwhelming your API or database.

> * Trigger multiple functions from a single event ([fan-out jobs](https://www.inngest.com/docs/guides/fan-out-jobs)).
> * [Delay code](https://www.inngest.com/docs/functions/multi-step) to run after a period of time.
> * [Debounce events](https://www.inngest.com/docs/reference/functions/debounce) to reduce duplicate processing.

Reviewer 2: ```suggestion
* Triggering multiple functions from a single event ([fan-out jobs](https://www.inngest.com/docs/guides/fan-out-jobs)).
* [Delaying code](https://www.inngest.com/docs/functions/multi-step) to run after a period of time.
* [Debouncing events](https://www.inngest.com/docs/reference/functions/debounce) to reduce duplicate processing.
```

In this guide, you'll learn how to set up Inngest to receive Clerk webhook events and how to define Inngest functions in your application using Clerk events including synchronizing data and sending welcome emails.

To follow this guide, you need an Inngest account and have Inngest [set up](https://www.inngest.com/docs/quick-start) in your codebase.

## Setting up the Inngest webhook

> To create an Inngest webhook endpoint and add it to your Clerk account, open the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) page at the
> 
> Clerk dashboard. Next, select the **Add Endpoint** button.

Reviewer 2: ```suggestion
To create an Inngest webhook endpoint and add it to your Clerk account, open the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) page at the Clerk dashboard. Next, select the **Add Endpoint** button.
```

<Images
  width={2578}
  height={1652}
  src="/docs/images/integrations/inngest/webhook-page.webp"
  alt="The Webhooks page in the Clerk Dashboard. A red arrow points to the button for Add Endpoint."
/>

> On the next page, select the **Transformation template** tab and the **Inngest** template, then click on the **Connect to Inngest **button**.**

Reviewer 2: ```suggestion
On the next page, select the **Transformation template** tab and the **Inngest** template, then click on the **Connect to Inngest** button.
```

   const user = event.data; // The event payload's data will be the Clerk User json object
   const { id, first_name, last_name } = user;
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email
   await database.users.insert({ id, email, first_name, last_name })
 }
)
```

The `event` object contains all of the relevant data for the event. The `event.data` will match the `data` object from the standard Clerk webhook [payload structure](/docs/integrations/webhooks#payload-structure). With this `clerk/user.created` event, the `event.data` will be a Clerk User json object.

As you can see, you can choose which events you want to handle with each function. You might write a separate function for `clerk/user.updated` and `clerk/user.deleted` handling the entire lifecycle end to end.

Note that multiple functions can also listen to the same event. This pattern is called “[fan-out](https://www.inngest.com/docs/guides/fan-out-jobs).”

## Creating a function to send a welcome email

Often, applications need to perform additional tasks when a new user is created, like send a welcome email with tips and useful information.

> While it is possible to add this logic at the end of your sync function as seen in the previous section, it's better to decouple logic into different functions so issues with one task do not affect the other ones. For example, if your email fails to send, it should not affect adding the user to an email list.

Reviewer 2: ```suggestion
While it is possible to add this logic at the end of your sync function as seen in the [previous section](/docs/integrations/webhooks/inngest#creating-a-function-to-sync-a-new-user-to-a-database), it's better to decouple logic into different functions so issues with one task do not affect the other ones. For example, if your email fails to send, it should not affect adding the user to an email list.
```

The code below creates another function using the same `clerk/user.created` event and adds the logic to send the welcome email:

> ```typescript

Reviewer: ```suggestion
```typescript  filename="src/inngest/welcome-emails.ts"
```

```typescript
const sendWelcomeEmail = inngest.createFunction(
 { id: 'send-welcome-email' },
 { event: 'clerk/user.created' },
 async ({ event }) => {
   const { user } = event.data
   const { first_name } = user
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email
   await emails.sendWelcomeEmail({ email, first_name })
 }
)
```

Now, you have a function that utilizes the same Clerk webhook event for another purpose. Clerk webhook events can be used for all sorts of application lifecycle use cases. For example, adding users to a marketing email list, starting a Stripe trial, or provisioning new account resources.

### Sending a delayed follow-up email

> Every Inngest function handler has an additional `step `object which provides tools to create more complex functions. Using `step.run` allows you to encapsulate specific code that will be automatically retried ensuring that issues with one part of your function don't force the entire function to re-run. Additionally, [other tools like `step.sleep`](https://www.inngest.com/docs/reference/functions/step-sleep), are available to extend functionality.

Reviewer 2: ```suggestion
Every Inngest function handler has an additional `step` object which provides tools to create more complex functions. Using `step.run` allows you to encapsulate specific code that will be automatically retried ensuring that issues with one part of your function don't force the entire function to re-run. Additionally, [other tools like `step.sleep`](https://www.inngest.com/docs/reference/functions/step-sleep), are available to extend functionality.
```

> The code below uses `step.sleep `to send a welcome email, then wait for three days, and follow up with a message offering a free trial:

Reviewer 2: ```suggestion
The code below uses `step.sleep `to send a welcome email, then wait for three days, and follows up with a message offering a free trial:
```

> ```typescript

Reviewer: ```suggestion
```typescript filename="src/inngest/welcome-emails.ts"
```

```typescript
const sendOnboardingEmails = inngest.createFunction(
 { id: 'onboarding-emails' },
 { event: 'clerk/user.created' },
 async ({ event, step }) => { // ← step is available in the handler's arguments
   const { user } = event.data
   const { first_name } = user
   const email = user.email_addresses.find(e =>
     e.id === user.primary_email_address_id
   ).email

   await step.run('welcome-email', async () => {
     await emails.sendWelcomeEmail({ email, first_name })
   })

   // wait 3 days before second email
   await step.sleep('wait-3-days', '3 days')

   await step.run('trial-offer-email', async () => {
     await emails.sendTrialOfferEmail({ email, first_name })
   })
 }
)
```

Now, you've extended the usefulness of Clerk webhook events even further to build an onboarding drip email campaign in just a few lines of code.

## Testing webhook events using the Inngest Dev Server

> During local development with Inngest, you can use the Inngest Dev Server to run and test your functions on your own machine. To start the server, in your project directory run the following command:

Reviewer 2: ```suggestion
During local development with Inngest, you can use the [Inngest Dev Server](https://www.inngest.com/docs/local-development) to run and test your functions on your own machine. To start the server, in your project directory run the following command:
```

```sh filename="terminal"
npx inngest-cli@latest dev
```

In your browser open [http://localhost:8288](http://localhost:8288) to see the development UI.

> To quickly get events to test within Dev Server, you can select any individual event from the Inngest Events tab then select the **Send to Dev Server**.

Reviewer 2: ```suggestion
To quickly get events to test within Dev Server, you can select any individual event from the **Events** tab then select the **Send to Dev Server**.
```

<Images
  width={2142}
  height={1246}
  src="/docs/images/integrations/inngest/inngest-send-to-dev-server.webp"
  alt="The Inngest Dashboard showing an individual event payload. Red arrows point to the Events tab and the Send to Dev Server button."
/>

> You'll now see the event in the Inngest Dev Server's stream tab alongside any functions that it triggered.

Reviewer 2: ```suggestion
You'll now see the event in the Inngest Dev Server's **Stream** tab alongside any functions that it triggered.
```
